File: examples/latest/lite/ios/CameraLite/CameraLite/CameraAnimator.swift
/*
 * Copyright (C) 2019-2021 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import heresdk
import UIKit

// A simple class that takes care of smooth map "moveTo"-animations
// using Apple's CADisplayLink.
class CameraAnimator {

    private let camera: CameraLite
    private var currentCamera: CameraUpdateLite!
    private var targetCamera: CameraUpdateLite!
    private var startTimeInSeconds: Double = 0
    private var previousTimeInSeconds: Double = 0
    private var animationDurationInSeconds: Double = 2

    private lazy var displayLink = CADisplayLink(target: self,
                                                 selector: #selector(animatorLoop))

    init(_ camera: CameraLite) {
        self.camera = camera
    }

    func setDurationInSeconds(animationDurationInSeconds: Double) {
        self.animationDurationInSeconds = animationDurationInSeconds
    }

    func moveTo(_ destination: GeoCoordinates, _ targetZoom: Double) {
        currentCamera = CameraUpdateLite(tilt: camera.getTilt(),
                                         bearing: camera.getBearing(),
                                         zoomLevel: camera.getZoomLevel(),
                                         target: camera.getTarget())

        // Take the shorter bearing difference.
        let targetBearing: Double = camera.getBearing() > 180 ? 360 : 0

        targetCamera = CameraUpdateLite(tilt: 0,
                                        bearing: targetBearing,
                                        zoomLevel: targetZoom,
                                        target: destination)

        // Start the run loop to execute animatorLoop() periodically.
        startTimeInSeconds = -1
        displayLink.isPaused = false
        displayLink.add(to: .current, forMode: .common)
    }

    @objc private func animatorLoop() {
        if startTimeInSeconds == -1 {
            // 1st frame, there's no previous frame.
            startTimeInSeconds = displayLink.timestamp
            previousTimeInSeconds = startTimeInSeconds
            return
        }

        let currentTimeInSeconds = displayLink.timestamp
        let elapsedTimeInSeconds = currentTimeInSeconds - startTimeInSeconds

        let frameDurationInSeconds = currentTimeInSeconds - previousTimeInSeconds
        let remainingFrames = (animationDurationInSeconds - elapsedTimeInSeconds) / frameDurationInSeconds

        // Calculate the new camera update.
        currentCamera = interpolate(currentCamera, targetCamera, remainingFrames)

        if elapsedTimeInSeconds >= animationDurationInSeconds {
            displayLink.isPaused = true
            camera.updateCamera(cameraUpdate: targetCamera)
            return
        }

        camera.updateCamera(cameraUpdate: currentCamera)
        previousTimeInSeconds = displayLink.timestamp
    }

    private func interpolate(_ currentCamera: CameraUpdateLite,
                             _ targetCamera: CameraUpdateLite,
                             _ remainingFrames: Double) -> CameraUpdateLite {
        let newTilt = interpolateLinear(currentValue: currentCamera.tilt,
                                        targetValue: targetCamera.tilt,
                                        remainingFrames: remainingFrames)

        let newBearing = interpolateLinear(currentValue: currentCamera.bearing,
                                           targetValue: targetCamera.bearing,
                                           remainingFrames: remainingFrames)

        let newZoomLevel = interpolateLinear(currentValue: currentCamera.zoomLevel,
                                             targetValue: targetCamera.zoomLevel,
                                             remainingFrames: remainingFrames)

        let newTargetLatitude = interpolateLinear(currentValue: currentCamera.target.latitude,
                                                  targetValue: targetCamera.target.latitude,
                                                  remainingFrames: remainingFrames)

        let newTargetLongitude = interpolateLinear(currentValue: currentCamera.target.longitude,
                                                   targetValue: targetCamera.target.longitude,
                                                   remainingFrames: remainingFrames)

        return CameraUpdateLite(tilt: newTilt,
                                bearing: newBearing,
                                zoomLevel: newZoomLevel,
                                target: GeoCoordinates(latitude: newTargetLatitude,
                                                       longitude: newTargetLongitude))
    }

    private func interpolateLinear(currentValue: Double,
                                   targetValue: Double,
                                   remainingFrames: Double) -> Double {
        let delta = (currentValue - targetValue) / remainingFrames
        let newValue = currentValue - delta

        // Overflow check.
        if (currentValue < targetValue) {
            if newValue >= targetValue {
                return targetValue
            }
        } else {
            if newValue <= targetValue {
                return targetValue
            }
        }

        return newValue
    }
}

File: examples/latest/lite/android/CameraLite/app/src/main/java/com/here/sdk/camera/CameraAnimator.java
 /*
  * Copyright (C) 2019-2022 HERE Europe B.V.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *     http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  *
  * SPDX-License-Identifier: Apache-2.0
  * License-Filename: LICENSE
  */

package com.here.sdk.camera;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.TimeInterpolator;
import android.animation.ValueAnimator;

import com.here.sdk.core.GeoCoordinates;
import com.here.sdk.mapviewlite.Camera;
import com.here.sdk.mapviewlite.CameraUpdate;

import java.util.ArrayList;
import java.util.List;

/**
 * A simple class that takes care of smooth map "moveTo"-animations
 * using Android's Animation framework.
 */
public class CameraAnimator {

    // Default animation should take 2 seconds.
    private static final int DURATION_IN_MILLIS = 2 * 1000;

    // Only one instance allowed.
    private static AnimatorSet animatorSet;

    private final Camera camera;
    private final List<Animator> valueAnimatorList = new ArrayList<>();
    private TimeInterpolator timeInterpolator;
    private int animationDurationInMillis = DURATION_IN_MILLIS;

    public CameraAnimator(Camera camera) {
        this.camera = camera;
    }

    public void setTimeInterpolator(TimeInterpolator timeInterpolator) {
        this.timeInterpolator = timeInterpolator;
    }

    public void setDurationInMillis(int animationDurationInMillis) {
        this.animationDurationInMillis = animationDurationInMillis;
    }

    public void moveTo(GeoCoordinates destination, double targetZoom) {
        CameraUpdate targetCameraUpdate = createTargetCameraUpdate(destination, targetZoom);
        createAnimation(targetCameraUpdate);
        startAnimation(targetCameraUpdate);
    }

    private CameraUpdate createTargetCameraUpdate(GeoCoordinates destination, double targetZoom) {
        double targetTilt = 0;

        // Take the shorter bearing difference.
        double targetBearing = camera.getBearing() > 180 ? 360 : 0;

        return new CameraUpdate(targetTilt, targetBearing, targetZoom, destination);
    }

    private void createAnimation(CameraUpdate cameraUpdate) {
        valueAnimatorList.clear();

        // Interpolate current values for zoom, tilt, bearing, lat/lon to the desired new values.
        ValueAnimator zoomValueAnimator = createAnimator(camera.getZoomLevel(), cameraUpdate.zoomLevel);
        ValueAnimator tiltValueAnimator = createAnimator(camera.getTilt(), cameraUpdate.tilt);
        ValueAnimator bearingValueAnimator = createAnimator(camera.getBearing(), cameraUpdate.bearing);
        ValueAnimator latitudeValueAnimator = createAnimator(
                camera.getTarget().latitude, cameraUpdate.target.latitude);
        ValueAnimator longitudeValueAnimator = createAnimator(
                camera.getTarget().longitude, cameraUpdate.target.longitude);

        valueAnimatorList.add(zoomValueAnimator);
        valueAnimatorList.add(tiltValueAnimator);
        valueAnimatorList.add(bearingValueAnimator);
        valueAnimatorList.add(latitudeValueAnimator);
        valueAnimatorList.add(longitudeValueAnimator);

        // Update all values together.
        longitudeValueAnimator.addUpdateListener(animation -> {
            float zoom = (float) zoomValueAnimator.getAnimatedValue();
            float tilt = (float) tiltValueAnimator.getAnimatedValue();
            float bearing = (float) bearingValueAnimator.getAnimatedValue();
            float latitude = (float) latitudeValueAnimator.getAnimatedValue();
            float longitude = (float) longitudeValueAnimator.getAnimatedValue();

            GeoCoordinates intermediateGeoCoordinates = new GeoCoordinates(latitude, longitude);
            camera.updateCamera(new CameraUpdate(tilt, bearing, zoom, intermediateGeoCoordinates));
        });
    }

    private ValueAnimator createAnimator(double from, double to) {
        ValueAnimator valueAnimator = ValueAnimator.ofFloat((float) from, (float )to);
        if (timeInterpolator != null) {
            valueAnimator.setInterpolator(timeInterpolator);
        }
        return valueAnimator;
    }

    private void startAnimation(CameraUpdate cameraUpdate) {
        if (animatorSet != null) {
            animatorSet.cancel();
        }

        animatorSet = new AnimatorSet();
        animatorSet.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                camera.updateCamera(cameraUpdate);
            }
        });

        animatorSet.playTogether(valueAnimatorList);
        animatorSet.setDuration(animationDurationInMillis);
        animatorSet.start();
    }
}
